CMakeCommon.cmake

```diff
@@ -221,6 +221,13 @@ function (GenerateAddOnProject target acVersion devKitDir addOnName addOnSources
     target_include_directories (${target} SYSTEM PUBLIC ${devKitDir}/Inc)
     target_include_directories (${target} PUBLIC ${addOnSourcesFolder})
 
+    # use GSRoot custom allocators consistently in the Add-On
+    target_precompile_headers(${target} 
+        PRIVATE 
+        ${devKitDir}/Modules/GSRoot/GSNew.hpp 
+        ${devKitDir}/Modules/GSRoot/GSMalloc.hpp
+        )
```

Comment:
Technically, the proper way to **replace** `::operator new` and `::operator delete` is by having those symbols provided to the linker, not at the source code level. `malloc` and friends cannot be overwritten in the same way.

The approach of these headers instead is to macro them out, which I do not find satisfying at all. However, I understand that something like a `GS::UniString` passed across a DLL boundary by value should be destructible in the caller DLL for convenience.

I understand that currently it is an observable property of precompiled headers that they are force included, but such a thing is merely an implementation detail and should not be relied on. Besides, `target_precompile_headers` is called wrong here. `devKitDir` is not guaranteed to be an absolute path and I would heavily recommend specifying headers here in a way they are expected to be used in code, because compilers can be very fussy when it comes to precompiled headers even when it's about `#include ""` vs `#include <>`.

I would suggest doing this instead:

```suggestion
    get_filename_component(new_hpp "${devKitDir}/Modules/GSRoot/GSNew.hpp" REALPATH)
    get_filename_component(malloc_hpp "${devKitDir}/Modules/GSRoot/GSMalloc.hpp" REALPATH)
    if(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
        target_compile_options(
            "${target}" PRIVATE
            "SHELL:/FI \"${new_hpp}\""
            "SHELL:/FI \"${malloc_hpp}\""
        )
    elseif(CMAKE_CXX_COMPILER_ID MATCHES Clang)
        target_compile_options(
            "${target}" PRIVATE
            "SHELL:-include \"${new_hpp}\""
            "SHELL:-include \"${malloc_hpp}\""
        )
    else()
        message(FATAL_ERROR "Unknown compiler ID. Please open an issue at https://github.com/GRAPHISOFT/archicad-addon-cmake-tools")
    endif()
```

Reply:
The current implementation is kind of proper, as compiling new/delete into a translation unit is a way of eventually providing it to the linker 😃.
But I get what you mean. Symbols exported from libraries would be preferable. But those also have their 
platform dependent quirks as library behavior is not standardized. AFAIK on Windows you have to link all libraries to your library containing the overridden new/delete, while on mac the loader takes care of any overridden weak symbols, and you can get a consistent override behavior in your application. Even in modules that weren't linked with the library implementing the override.
@peter-esik might know why the GSRoot implementation is the way it is, but changing the way it works is really not an option at this point.

I agree that force includes and precompiled headers are not the same, and precompiled headers are not standardized, so we can't guarantee that they are substitutes for force includes. 
But the observable behavior is indeed that they behave as force includes across compiler implementations. This behavior is unlikely to change, as it would be a compatibility breaking change that compiler vendors usually try to avoid. I also don't think there's a real alternative to the current behavior of precompiled headers.
I would've went with force includes if CMake provided an abstraction for them, but as they didn't, target_precompile_headers seemed to be the next best thing.
I think both my solution and your suggestion is the same in the sense that they're both rely on compiler specific behavior. While mine is semantically less correct, the comment communicates the intent (functional reason, not performance). My solution has a theoretical chance of breaking in the future, while it has a high change of working with other compilers. It's also a bit simpler FWIW 😃. Please help if I missed some comparison aspects.

Regarding absolute paths: Wouldn't it be better to call RELPATH on devKitDir, to make sure it's an absolute path everywhere? (I don't expect the appended parts to include symlinks.) Or is absolute path only problematic for header paths?

Reply:
I think the "override" for global new and delete is like this because back at the time this was the easiest solution. The C++ standard conforming way would be to provide these symbols to the linker in each module. This, of course, must happen in Add-Ons, 3rd party modules, etc.

In other words, we should create a static library containing these symbols, and link this to all modules. Roland had a solution for this up for [review](https://fisheye.graphisoft.hu/cru/GEN-31654), but this has not been sumbitted/finished.

An interesting note: Chrome deals with this by fiddling with their binaries _after_ they are produced, and overwrite `malloc` and `free` in them after they are linked. This is made possible because they use static linking for everything.

Reply:
It's not this PR's job to fix the replacement of those operators, but the way things are belong in the less-than-ideal category.

Using `REALPATH` is just my preference.

On a second thought, even this solution might not be sufficient. What about shared library dependencies of addons built using this template? Is this requirement documented?  
Otherwise, please don't (ab)use the precompiled header feature for this and stick to the suggested code. If you think I missed additional compilers, please adjust accordingly using [CMAKE_CXX_COMPILER_ID](https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_COMPILER_ID.html).